miio/integrations/ihcooker/tests/test_ihcooker.py

```diff
@@ -0,0 +1,321 @@
+            "get_prop": self._get_state,
+            "set_start": lambda profile: self._validate_profile(profile[0]),
+            "set_menu1": lambda profile: self._validate_profile(profile[0]),
+        }
+        super().__init__(args, kwargs)
+
+    def _get_state(self, vars):
+        """Return wanted properties."""
+        assert vars == ["all"], "Only get_props all is supported"
+        return self.state
+
+    @staticmethod
+    def _validate_profile(profile):
+        """Validates if profile is a valid string."""
+        assert all(c in string.hexdigits for c in profile) and len(profile) == 358
+
+
+class DummyIHCookerV1(DummyIHCookerV2):
+    def __init__(self, *args, **kwargs):
+        super().__init__(args, kwargs)
+        self._model = MODEL_V1
+
+        self.state = [
+            "waiting",
+            "02537465616d2f626f696c000000000002",
+            "0013000000",
+            "0100010000000000",
+            "001b0a04",
+            "00000001000000020000000300000004000000050000a4e000000dcc00000000",
+            "00",
+            "0000e60a0017130f00",
+        ]
+
+
+@pytest.fixture(scope="class")
+def ihcooker_exp1(request):
+    request.cls.device = DummyIHCookerV2()
+
+
+@pytest.fixture(scope="class")
+def ihcooker_v1(request):
+    request.cls.device = DummyIHCookerV1()
+
+
+@pytest.mark.usefixtures("ihcooker_exp1")
+class TestIHCookerV2(TestCase):
+    def state(self):
+        return self.device.status()
+
+    def test_recipe(self):
+        recipe = {}
+        self.device.start(recipe)
+
+    def test_set_menu(self):
+        recipe = {}
+        self.device.set_menu(recipe, 1)
+
+    def test_start_temp(self):
+        self.device.start_temp(temperature=30, minutes=30)
+
+    def test_start_json(self):
+        json_str = """{
+  "menu_location": 5,
+  "recipe_name": "Rice Cooking",
+  "recipe_id": 42,
+  "duration_minutes": 300,
+  "stages": [
+    {
+      "mode": "Unknown10",
+      "temp_threshold": 60,
+      "temp_target": 90,
+      "power": 99
+    },
+    {
+      "mode": "Unknown10",
+      "temp_threshold": 91,
+      "temp_target": 102,
+      "power": 10
+    },
+    {
+      "mode": "TempAutoSmallPot",
+      "minutes": 300,
+      "temp_threshold": 150,
+      "temp_target": 60,
+      "power": 10
+    }
+  ]
+}
+"""
+        self.device.start(json_str)
+
+    def test_construct(self):
+        recipe = (
+            "030405546573740a52656369706500000000000000000000000000000000000000000003ea0000000"
+            "000000000000100008005323c111113028006333d121012188007343e130f11000000f93e00141400"
+            "0000f93e001414000000f93e001414000000f93e001414000000f93e001414000000f93e001414000"
+            "000f93e001414000000f93e001414000000f93e001414000000f93e001414000000f93e0014140000"
+            "00f93e001414000000000000000000df87"
+        )
+
+        res = profile_v2.parse(bytes.fromhex(recipe))
+        self.assertEqual(profile_v2.parse(profile_v2.build(res)), res)
```

Comment:
Done! What is the reason for this?

Reply:
The use of testcase classes is a remnant from unittest's java roots. The preferred way nowadays is to avoid that class altogether and use individual test functions, or that's my understanding at least.

One of the reasons is that it's also cleaner and feels more idiomatic python, as you don't need to think about which [assert function](https://docs.python.org/3/library/unittest.html#assert-methods) should be used.